tests/integration/test_alerts.py

```diff
@@ -47,6 +49,23 @@
         ALERT_ADVANCED_QUERY_COMMAND,
     ],
 )
-def test_alert_command_returns_success_return_code(command, command_runner):
-    return_code, response = command_runner(command)
-    assert return_code == 0
+def test_alert_command_returns_success_return_code(
+    runner, integration_test_profile, command
+):
+    assert_test(runner, integration_test_profile, command)
+
+
+@pytest.mark.integration
+@pytest.mark.parametrize(
```

Comment:
You tend to use parameterized tests a lot in a way that is difficult to read or change (as I experienced on Monday.

Parameterized tests should only be used when there is a simple parameter.  Whole complex arg strings are not good candidates here. It would be better to have repetitive individual tests than to have these parameterized tests everywhere

Reply:
the command string is the simplest parameter, with each line one can see all the different possible command combinations instead of reading each method. I don't know if I can have an explanation why this is easier for me.

Reply:
To convince team to allow this, when ssl is ready all we need in integration is to add one single line with the command in parametrize block and that will be it. ( given that the `DataServer` class is sufficient )

Reply:
If a test fails, there is no way to run just that test. You have to run them all. This is not how parameterized tests should be used. We have been doing this a lot in the integration tests. It also makes it difficult to know which tests are still needed and which ones we have. A command string is a string of args -- not a single arg.

Reply:
So, if the SSL test fails, we now have to run all the other integration tests each time or comment them out. 

 Also, when looking at the Test Failure report, it is not immediately clear which of the parameterized tests was the one that failed.

Reply:
I agree that testing individual test with parameterize is difficult, there is a plugin that can be used to execute only failed tests from previous execution, but overall I see the pain here.
However on the second concern, the report does give exactly which test failed. In fact with parameterize we can see the command in the log against the failure in the report. Which we can use directly without a need to open the test file and see what command was executed under that method.

Reply:
This is the log I see when test is executed. Observe how command being passed as in input from parameterize is mentioned in it, the `-TCP` is another parameter that is passed for other purpose.  
```
$ pytest -s -x -v tests -m integration -k send-to
collected 677 items / 671 deselected / 6 selected                                                                                                                                                          

tests/integration/test_alerts.py::test_alerts_send_to[alerts send-to localhost:5140 -p TCP -b 2021-01-09-TCP] PASSED
tests/integration/test_alerts.py::test_alerts_send_to[alerts send-to localhost:5140 -p UDP -b 2021-01-09-UDP] PASSED
tests/integration/test_auditlogs.py::test_auditlogs_send_to[audit-logs send-to localhost:5140 -p TCP -b '2021-01-27 13:52:56'-TCP]
```

Reply:
Seems like you might have tried `pytest` without `-v` option and hence all the confusion!! In tox we do use `-v` option so this shouldn't be a concern.

Reply:
When a test fails for me, I typically use Pycharm, and that is where the output seems confusing to me, especially when there are lots of different commands in the same parameter.

This is just simply an abuse of parameterized tests.  Technically, you can put every single test in the whole suite into one giant parameterized test. Don't do this please!

Also, this is just part of the team process that has been discussed in the past on other occasions, ask @alanag13 for confirmation there.

Using parameterized tests this way increases the complexity of running the tests, and in make-or-break moments (such as the Monday of the latest release), it is frustrating to try and sift through these.  I hope you understand.

I will approve for now, but going forward, please let's not do this.

Reply:
There is also the point that is just hard to know what tests we have and which ones are missing, when looking at the code. We have to parse a python dictionary with our eyeballs

Reply:
I would happily change this if everyone else feels it so. However, I would share my disagreement with above raised concerns, particularly that "using parameterize here is abuse of the feature" 

- As per the [doc](https://docs.pytest.org/en/stable/example/parametrize.html#generating-parameters-combinations-depending-on-command-line)  the usage of `parameterize` is to provide all the possible combinations in a single place. In the current implementation I have not created a single test case to cover everything,  `send-to` and `list` commands have separate test methods and then all combinations of each sub-command are parameterized.

- On "usage of `pycharm`", I feel there has to be a way to configure it to pass other options like  `-v`, `-x` which would be helpful during debugging.

- On "team process", I would like to say, the context has changed and so team needs to reconsider the process. In unit tests' context it might (not sure) make sense to parameterize but in integration tests where there are multiple combinations to test we have to reconsider it.

- On "missing tests", the number of combinations possible in `list` commands are exponentially high and cannot all be tested, I have just made sure I cover  each option at-least once. Whether writing a method for each separate combination will help to figure out  missing tests is subjective and depends on reader.

I am ready to rework and write a separate method for each command if someone else is going to rework on this again, its better I do it. imo, its wastage of resources time and under utilisation of tools. I feel `parameterize` is a good feature of the tool `pycharm` and we should be ok using it. 

@alanag13  @timabrmsn thoughts!!

Reply:
I think we're just pushing too much shared info into the parameters, since we're passing the whole CLI cmd string instead of just the changing parameters for each run. 

We should turn this:

```
@pytest.mark.integration
@pytest.mark.parametrize(
    "command",
    [
        "{} list".format(ALERT_RULES_COMMAND),
        "{} list -f CSV".format(ALERT_RULES_COMMAND),
        "{} list -f TABLE".format(ALERT_RULES_COMMAND),
        "{} list -f RAW-JSON".format(ALERT_RULES_COMMAND),
        "{} list -f JSON".format(ALERT_RULES_COMMAND),
        "{} list --format CSV".format(ALERT_RULES_COMMAND),
        "{} list --format TABLE".format(ALERT_RULES_COMMAND),
        "{} list --format JSON".format(ALERT_RULES_COMMAND),
        "{} list --format RAW-JSON".format(ALERT_RULES_COMMAND),
    ],
)
def test_alert_rules_list_command_returns_success_return_code(
    runner, integration_test_profile, command
):

    assert_test_is_successful(runner, append_profile(command))
```

Into something like this:

```
@pytest.mark.integration
@pytest.mark.parametrize(
    "options",
    [
        [],
        ["-f", "CSV"],
        ["-f", "TABLE"],
        ["-f", "RAW-JSON"],
        ["-f", "JSON"],
        ["--format", "CSV"],
        ["--format", "TABLE"],
        ["--format", "JSON"],
        ["--format", "RAW-JSON"],
    ],
)
def test_alert_rules_list_command_returns_success_return_code(
    runner, integration_test_profile, options
):
    cmd = ["alert-rules", "list", "--profile", "TEMP-INTEGRATION-TEST", *options]
    assert_test_is_successful(runner, cmd)
```

It does the same exact work, it's just easier to isolate/update what's being parametrized vs what is static to this individual test.

Reply:
@timabrmsn This is certainly an improvement, and for most of the parameterized tests, this would be an improvement.

Reply:
Ok, I will go ahead and do the changes as requested, also is splitting the command into a list better, I prefer reading without quotes and commas but can cover that as well if that is important!!

Reply:
I'm ok with the cmd and options being strings that get combined then split when passed to the runner.

Reply:
@alanag13 said he would add his two cents here

Reply:
I have a couple thoughts here -- I see that I was @mentioned a couple times here
- first off, when the team disagrees on implementation details like this, it is probably best to simply note that further discussion is needed, schedule some time to agree on an approach, etc, but to not allow a disagreement like this to delay a release/feature merge.

That being said, If I could make a style rule that would emit a warning whenever a parameterized test was used, I probably would. Parameterized tests can be useful, but are often used for situations that can later cause difficulty. I find them most useful when the parameter list is very small, as a means to test different types of an input (will this work with a very small number? a very low number? a decimal number) that are all taking a similar code path. An example in this scenario might be to parameterize both `-f` and `--format` to make sure they both work as input. 

The biggest disadvantage is that parameterized tests cannot typically be run individually. This is not just an inconvenience -- I have seen this lead to nasty, difficult-to-diagnose bugs  and/or false positives in tests because a test was dependent on a previous parameter running first.

Additionally, especially as someone who works in multiple frameworks/languages, parameterized tests introduce _uncertainty_ to me:
- if one of the parameterized tests fails, will I know which one?
- if one of the parameterized tests fails, do the rest still get run, or or does the test stop?
- if one of the parameterized tests fails, will the problem be in the same general section of code as the other parameters?

I have worked with testing tools where some of the answers to these questions is _no_. Even if the answer to all of these is _yes_, the fact that I do not know right away just by looking at the code is off-putting. This uncertainty doesn't occur when using one function per test, which is why I typically avoid them.

Typically, if these were unit tests, I would argue that `CSV/TABLE/RAW-JSON/JSON` takes significantly different code paths depending on which you choose and should be separate tests, but in this case (since it is an integration test), we can be reasonably confident that the process that thing we're trying to test works as expected using only one of those parameters -- there may not even be a need to create a separate test for each one. Recall that the point of the integration test is essentially just to test that we can hit the network successfully. Anything that could potentially go wrong due to your choice of format should be covered by a unit test.

Reply:
To summarize -- In the end, whether or not to use parameterized tests can probably be categorized as a personal preference. I have had experiences that have given me a strong personal preference to avoid them which i'm happy to share to allow others to avoid the same pain -- but I don't want this project to be built solely around enforcing my personal preferences.

Reply:
Made changes based on above discussion.

Committed the changes in two steps, i.e a commit each for the step as follows:

1. Made changes as suggested by Tim to pass single option in parameterize
2. Removed parameterize options all together based on the point that  _" the point of the integration test is essentially just to test that we can hit the network successfully in each common use-case. Anything that could potentially go wrong due to your choice of format should be covered by a unit test."_

Reply:
If someone wants to have all the tests back, I can easily do it by getting rid of the last commit.

Reply:
For `send-to` sub-command tests, parameterize still exists but now it accepts just the protocol field as option which has just two values.

Reply:
To answer Alan's queries above, the answer is yes to all three of them.